numpy/core/src/private/ufunc_override.c

```diff
@@ -67,68 +89,61 @@ PyUFunc_WithOverride(PyObject *args, PyObject *kwds,
                      PyObject **with_override, PyObject **methods)
 {
     int i;
-
-    int nargs;
-    int nout_kwd = 0;
-    int out_kwd_is_tuple = 0;
     int num_override_args = 0;
-
+    int nin, nout = 0;
     PyObject *out_kwd_obj = NULL;
-    /*
-     * Check inputs
-     */
-    if (!PyTuple_Check(args)) {
-        PyErr_SetString(PyExc_TypeError,
-                        "Internal Numpy error: call to PyUFunc_HasOverride "
-                        "with non-tuple");
-        goto fail;
-    }
-    nargs = PyTuple_GET_SIZE(args);
-    if (nargs > NPY_MAXARGS) {
-        PyErr_SetString(PyExc_TypeError,
-                        "Internal Numpy error: too many arguments in call "
-                        "to PyUFunc_HasOverride");
-        goto fail;
-    }
-    /* be sure to include possible 'out' keyword argument. */
-    if (kwds && PyDict_CheckExact(kwds)) {
+
+    nin = PyTuple_Size(args);
+    if (nin < 0) {
+        return -1;
+    }
+    if (kwds == NULL) {
+        nout = 0;
+    }
+    else {
+        if (!PyDict_CheckExact(kwds)) {
+            PyErr_SetString(PyExc_TypeError,
+                        "Internal Numpy error: call to PyUFunc_WithOverride "
+                        "with non-dict kwds");
+            return -1;
+        }
         out_kwd_obj = PyDict_GetItemString(kwds, "out");
-        if (out_kwd_obj != NULL) {
-            out_kwd_is_tuple = PyTuple_CheckExact(out_kwd_obj);
-            if (out_kwd_is_tuple) {
-                nout_kwd = PyTuple_GET_SIZE(out_kwd_obj);
-            }
-            else {
-                nout_kwd = 1;
+        if (out_kwd_obj == NULL) {
+            nout = 0;
+        }
+        else if (PyTuple_CheckExact(out_kwd_obj)) {
+            nout = PyTuple_GET_SIZE(out_kwd_obj);
+            if (nout == 1) {
+                /* saves special-casing the tuple-of-one below */
+                out_kwd_obj = PyTuple_GET_ITEM(out_kwd_obj, 0);
             }
         }
+        else {
+            nout = 1;
```

Comment:
Wouldn't it be better to just cast to a 1-tuple here? We already do that when we call `__array_wrap__` and `__array_ufunc__`

Reply:
Could also do:

```
PyObject **out_objs;
if (out_kwd_obj == NULL) {
    nout = 0;
    out_objs = NULL;
}
else if (PyTuple_CheckExact(out_kwd_obj)) {
    nout = PyTuple_GET_SIZE(out_kwd_obj)
    out_objs = PySequence_Fast_ITEMS(out_kwd_obj);
}
else {
    nout = 1
    out_objs = &out_kwd_obj;
}


// Cleanup just xdecrefs out_kwd_obj
``

Reply:
Could wrap this into

```
int parse_out_args(PyObject *args, PyObject *kwargs, PyObject **out_ref, PyObject ***out_objs) {
    // code above
    return nout
}
```

Which would at least prevent you repeating this logic twice in this file. We can solve the repetitions across files later.

Reply:
I like that logic very much, and have adapted it; in the end, it should percolate up.